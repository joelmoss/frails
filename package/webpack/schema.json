{
  "type": "object",
  "properties": {
    "devServer": {
      "anyOf": [
        {
          "description": "Enabled in development mode, but can be enabled/disabled completely by setting to `true`/`false`",
          "type": "boolean"
        },
        { "type": "object" }
      ]
    },
    "sideLoadViews": {
      "description": "Side load views with Javascript and CSS assets. Enabled by default (`true`).",
      "type": "boolean"
    }
  }
}
