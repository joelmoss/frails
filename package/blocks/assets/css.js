const MiniCssExtractPlugin = require("mini-css-extract-plugin");

const useCss = require("../../utils/use_css");

module.exports = css;

function css() {
  return (context, { addLoader, addPlugin }) => (prevConfig) => {
    const nextConfig = addLoader(
      Object.assign(
        {
          test: /\.css$/i,
          use: useCss({
            mode: context.mode,
            useModules: "auto",
            postcss: context.postcss,
          }),
        },
        context.match
      )
    )(prevConfig);

    return addPlugin(new MiniCssExtractPlugin())(nextConfig);
  };
}
