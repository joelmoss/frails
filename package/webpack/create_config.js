const path = require('path')
const debug = require('debug')('frails')
const { default: merge } = require('webpack-merge')
const validate = require('schema-utils')
const webpack = require('webpack')
const WebpackAssetsManifest = require('webpack-assets-manifest')
const FixStyleOnlyEntriesPlugin = require('webpack-fix-style-only-entries')

const frailsConfig = require('../config')
const applyDevServer = require('./config/dev_server')
const applySideLoadAssets = require('./config/side_load_views')
const schema = require('./schema.json')

module.exports = (options = {}) => {
  validate(schema, options)

  // Make sure NODE_ENV env variable is set.
  process.env.NODE_ENV = frailsConfig.railsEnv === 'test' ? 'development' : frailsConfig.railsEnv

  const config = merge(
    // Default config.
    {
      mode: process.env.NODE_ENV,
      entry: {},

      // Merge with user-provided options.
      ...options
    },

    // Required config - this should be untouched.
    {
      output: {
        path: frailsConfig.absolutePublicPath,
        publicPath: `/${frailsConfig.publicOutputPath}/`
      },
      resolve: {
        alias: {
          assets: path.resolve(frailsConfig.appPath, 'assets')
        }
      },
      plugins: [
        new FixStyleOnlyEntriesPlugin(),
        new WebpackAssetsManifest({
          writeToDisk: true,
          entrypoints: true,
          sortManifest: false,
          publicPath: true
        })
      ]
    }
  )

  applyDevServer(config)
  applySideLoadAssets(config)

  debug(config)

  return config
}
