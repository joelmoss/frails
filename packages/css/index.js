const { merge, mergeWithCustomize, unique } = require('webpack-merge')
const MiniCssExtractPlugin = require('mini-css-extract-plugin')
const { config: frailsConfig, createConfigBlock } = require('@frails/core')

module.exports = createConfigBlock('css', (options, baseConfig) => {
  if (options === false) return

  baseConfig.plugins = mergeWithCustomize({
    customizeArray: unique(
      'plugins',
      ['MiniCssExtractPlugin'],
      plugin => plugin.constructor && plugin.constructor.name
    )
  })(
    {
      plugins: baseConfig.plugins
    },
    {
      plugins: [new MiniCssExtractPlugin()]
    }
  ).plugins

  baseConfig.module = merge(baseConfig.module || {}, {
    rules: [
      {
        test: /\.css$/i,
        use: [
          'mini-css-extract-plugin',
          {
            loader: 'css-loader',
            options: {
              modules: {
                localIdentName: frailsConfig.cssLocalIdentName
              }
            }
          }
        ]
      }
    ]
  })
})
