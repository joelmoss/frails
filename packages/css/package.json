{
  "name": "@frails/css",
  "version": "0.9.0",
  "description": "Frails CSS configuration.",
  "repository": "git@github.com:joelmoss/frails.git",
  "bugs": "https://github.com/joelmoss/frails/issues",
  "author": "Joel Moss <joel@developwithstyle.com>",
  "license": "MIT",
  "keywords": [
    "webpack",
    "rails"
  ],
  "peerDependencies": {
    "webpack": "^4.0.0"
  },
  "dependencies": {
    "@frails/core": "^0.9.0",
    "mini-css-extract-plugin": "^0.11.2",
    "webpack-merge": "^5.1.4"
  }
}
