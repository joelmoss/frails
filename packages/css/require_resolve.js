// Only here tio allow us to mock `require.resolve`, which prevents snapshots from including the
// Yarn PnP virtual cache paths when resolving loaders.
module.exports = loader => require.resolve(loader)
