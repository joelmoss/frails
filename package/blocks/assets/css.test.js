import { createConfig } from "@webpack-blocks/webpack";

import css from "./css";

describe("assets/css", () => {
  test("config", () => {
    expect(createConfig(css())).toMatchSnapshot();
  });
});
