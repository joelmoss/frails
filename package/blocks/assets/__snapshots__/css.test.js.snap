// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assets/css config 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\\\\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "/Users/joelmoss/dev/frails/node_modules/mini-css-extract-plugin/dist/loader.js",
            "options": Object {
              "esModule": true,
            },
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "modules": Object {
                "auto": true,
                "localIdentName": "[local]-[md5:hash:hex:6]",
              },
              "sourceMap": false,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    MiniCssExtractPlugin {
      "options": Object {
        "chunkFilename": "[name].css",
        "filename": "[name].css",
        "ignoreOrder": false,
        "moduleFilename": [Function],
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".json",
    ],
  },
  "stats": Object {
    "children": false,
    "chunks": false,
    "modules": false,
    "reasons": false,
  },
}
`;
